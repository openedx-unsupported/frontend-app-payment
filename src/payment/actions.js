import { utils } from '../common';

const { AsyncActionType } = utils;

export const FETCH_BASKET = new AsyncActionType('PAYMENT', 'FETCH_BASKET');


export const fetchBasket = () => ({
  type: FETCH_BASKET.BASE,
  payload: {},
});

export const fetchBasketBegin = () => ({
  type: FETCH_BASKET.BEGIN,
});

export const fetchBasketSuccess = result => ({
  type: FETCH_BASKET.SUCCESS,
  payload: result,
});

export const fetchBasketFailure = () => ({
  type: FETCH_BASKET.FAILURE,
});
