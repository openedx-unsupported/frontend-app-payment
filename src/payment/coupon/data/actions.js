import { AsyncActionType } from '../../../common/utils';

export const ADD_COUPON = new AsyncActionType('PAYMENT', 'ADD_COUPON');
export const REMOVE_COUPON = new AsyncActionType('PAYMENT', 'REMOVE_COUPON');
export const UPDATE_COUPON_DRAFT = 'PAYMENT__UPDATE_COUPON_DRAFT';

// ADD COUPON

export const addCoupon = code => ({
  type: ADD_COUPON.BASE,
  payload: {
    code,
  },
});

export const addCouponBegin = () => ({
  type: ADD_COUPON.BEGIN,
});

export const addCouponSuccess = (voucherId, code, benefit) => ({
  type: ADD_COUPON.SUCCESS,
  payload: {
    benefit,
    code,
    voucherId,
  },
});

export const addCouponFailure = errorCode => ({
  type: ADD_COUPON.FAILURE,
  payload: {
    errorCode,
  },
});

// REMOVE COUPON

export const removeCoupon = voucherId => ({
  type: REMOVE_COUPON.BASE,
  payload: {
    voucherId,
  },
});

export const removeCouponBegin = () => ({
  type: REMOVE_COUPON.BEGIN,
});

export const removeCouponSuccess = result => ({
  type: REMOVE_COUPON.SUCCESS,
  payload: {
    result,
  },
});

export const removeCouponFailure = errorCode => ({
  type: REMOVE_COUPON.FAILURE,
  payload: {
    errorCode,
  },
});

// UPDATE COUPON DRAFT

export const updateCouponDraft = code => ({
  type: UPDATE_COUPON_DRAFT,
  payload: {
    code,
  },
});
