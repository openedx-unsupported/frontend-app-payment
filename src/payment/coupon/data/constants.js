export const PERCENTAGE_BENEFIT = 'Percentage';
export const ABSOLUTE_BENEFIT = 'Absolute';
