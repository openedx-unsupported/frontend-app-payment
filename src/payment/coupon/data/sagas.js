import { call, put, takeEvery } from 'redux-saga/effects';
import {
  ADD_COUPON,
  addCouponBegin,
  addCouponSuccess,
  addCouponFailure,
  removeCouponBegin,
  removeCouponSuccess,
  removeCouponFailure,
  REMOVE_COUPON,
} from './actions';
import { postCoupon } from './service';

function* handleAddCoupon(action) {
  yield put(addCouponBegin());
  try {
    const result = yield call(postCoupon, action.payload.code);
    const { id: voucherId, code, benefit } = result.voucher;
    yield put(addCouponSuccess(voucherId, code, benefit));
  } catch (e) {
    if (e.code !== undefined) {
      yield put(addCouponFailure(e.code));
    } else {
      throw e;
    }
  }
}

function* handleRemoveCoupon(action) {
  yield put(removeCouponBegin());
  try {
    const result = yield call(postCoupon, action.payload.codel);
    yield put(removeCouponSuccess(result));
  } catch (e) {
    if (e.code !== undefined) {
      removeCouponFailure(e.code);
    } else {
      throw e;
    }
  }
}

export default function* saga() {
  yield takeEvery(ADD_COUPON.BASE, handleAddCoupon);
  yield takeEvery(REMOVE_COUPON.BASE, handleRemoveCoupon);
}
