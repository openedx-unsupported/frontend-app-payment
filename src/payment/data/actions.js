import { createRoutine } from 'redux-saga-routines';

// Routines are action + action creator pairs in a series.
// Actions adhere to the flux standard action format.
// Routines by default are in the form of:
//
// Action                |   Action Creator
// -----------------------------------------------
// fetchBasket.TRIGGER   |   fetchBasket()
// fetchBasket.SUCCESS   |   fetchBasket.success()
// fetchBasket.FAILURE   |   fetchBasket.failure()
// fetchBasket.FULFILL   |   fetchBasket.fulfill()
//
// Created with redux-saga-routines
export const fetchCaptureKey = createRoutine('FETCH_CAPTURE_KEY');
export const submitPayment = createRoutine('SUBMIT_PAYMENT');
export const fetchBasket = createRoutine('FETCH_BASKET');
export const addCoupon = createRoutine('ADD_COUPON');
export const removeCoupon = createRoutine('REMOVE_COUPON');
export const updateQuantity = createRoutine('UPDATE_QUANTITY');

// Actions and their action creators
export const BASKET_DATA_RECEIVED = 'BASKET_DATA_RECEIVED';

export const basketDataReceived = basket => ({
  type: BASKET_DATA_RECEIVED,
  payload: basket,
});

export const BASKET_PROCESSING = 'BASKET_PROCESSING';

export const basketProcessing = isProcessing => ({
  type: BASKET_PROCESSING,
  payload: isProcessing,
});

export const CAPTURE_KEY_PROCESSING = 'CAPTURE_KEY_PROCESSING';

export const captureKeyProcessing = isProcessing => ({
  type: CAPTURE_KEY_PROCESSING,
  payload: isProcessing,
});

export const MICROFORM_STATUS = 'MICROFORM_STATUS';

export const microformStatus = status => ({
  type: MICROFORM_STATUS,
  payload: status,
});

export const CAPTURE_KEY_START_TIMEOUT = 'CAPTURE_KEY_START_TIMEOUT';

export const captureKeyStartTimeout = () => ({
  type: CAPTURE_KEY_START_TIMEOUT,
});
export const CAPTURE_KEY_DATA_RECEIVED = 'CAPTURE_KEY_DATA_RECEIVED';

export const captureKeyDataReceived = captureKey => ({
  type: CAPTURE_KEY_DATA_RECEIVED,
  payload: captureKey,
});
