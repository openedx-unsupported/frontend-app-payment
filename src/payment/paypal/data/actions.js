import { utils } from '../../../common';

const { AsyncActionType } = utils;

export const SUBMIT_PAYMENT_PAYPAL = new AsyncActionType('PAYMENT', 'SUBMIT_PAYMENT_PAYPAL');

export const submitPaymentPayPal = () => ({
  type: SUBMIT_PAYMENT_PAYPAL.BASE,
});

export const submitPaymentPayPalBegin = () => ({
  type: SUBMIT_PAYMENT_PAYPAL.BEGIN,
});

export const submitPaymentPayPalFailure = () => ({
  type: SUBMIT_PAYMENT_PAYPAL.FAILURE,
});
