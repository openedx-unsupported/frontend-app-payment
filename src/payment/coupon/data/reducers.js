import { ADD_COUPON, REMOVE_COUPON, UPDATE_COUPON_DRAFT } from './actions';

const defaultState = {
  benefit: null,
  code: null,
  voucherId: null,
  error: null,
  loaded: false,
  loading: false,
};

const reducer = (state = defaultState, action = null) => {
  if (action !== null) {
    switch (action.type) {
      case ADD_COUPON.BEGIN:
        return {
          ...state,
          loading: true,
          loaded: false,
          error: null,
        };
      case ADD_COUPON.SUCCESS:
        return {
          ...state,
          loading: false,
          loaded: true,
          error: null,
          benefit: action.payload.benefit,
          code: action.payload.code,
          voucherId: action.payload.voucherId,
        };
      case ADD_COUPON.FAILURE:
        return {
          ...state,
          loading: false,
          loaded: false,
          error: action.payload.error,
          voucherId: null,
        };
      case REMOVE_COUPON.BEGIN:
        return {
          ...state,
          loading: true,
          loaded: false,
          error: null,
        };
      case REMOVE_COUPON.SUCCESS:
        return defaultState;
      case REMOVE_COUPON.FAILURE:
        return {
          ...state,
          loading: false,
          loaded: true, // There's still a loaded coupon.
          error: action.payload.error,
        };
      case UPDATE_COUPON_DRAFT:
        return {
          ...state,
          code: action.payload.code,
        };
      default:
    }
  }

  return state;
};

export default reducer;
