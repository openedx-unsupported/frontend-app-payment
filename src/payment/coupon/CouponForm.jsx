import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Button, Input, ValidationFormGroup } from '@edx/paragon';
import { injectIntl, intlShape } from '@edx/frontend-i18n';

import messages from './messages';
import { addCoupon, removeCoupon, updateCouponDraft } from './data/actions';
import { PERCENTAGE_BENEFIT, ABSOLUTE_BENEFIT } from './data/constants';
import Benefit from './Benefit';

export class CouponForm extends Component {
  constructor(props) {
    super(props);

    this.handleChange = this.handleChange.bind(this);
    this.handleAddSubmit = this.handleAddSubmit.bind(this);
    this.handleRemoveSubmit = this.handleRemoveSubmit.bind(this);
  }

  handleChange(event) {
    const { value } = event.target;
    this.props.updateCouponDraft(value);
  }

  handleAddSubmit(event) {
    event.preventDefault();
    this.props.addCoupon(this.props.code);
  }

  handleRemoveSubmit(event) {
    event.preventDefault();
    this.props.removeCoupon(this.props.voucherId);
  }

  renderAdd() {
    const {
      code, intl, error, loading,
    } = this.props;

    const id = 'couponField';

    return (
      <form onSubmit={this.handleAddSubmit}>
        <ValidationFormGroup for={id} invalid={error !== null} invalidMessage={error}>
          <label className="h6 d-block" htmlFor={id}>
            {intl.formatMessage(messages['payment.coupon.label'])}
          </label>
          <Input
            className="mb-2"
            name={id}
            id={id}
            type="text"
            value={code || ''}
            onChange={this.handleChange}
          />
          <Button disabled={loading} className="btn-primary" type="submit">
            {intl.formatMessage(messages['payment.coupon.submit'])}
          </Button>
        </ValidationFormGroup>
      </form>
    );
  }

  renderRemove() {
    const { intl } = this.props;
    return (
      <form onSubmit={this.handleRemoveSubmit} className="d-flex align-items-center mb-3">
        {this.props.benefit !== null ? (
          <Benefit code={this.props.code} {...this.props.benefit} />
        ) : null}
        <Button className="btn-link display-inline p-0 pl-3 border-0" type="submit">
          {intl.formatMessage(messages['payment.coupon.remove'])}
        </Button>
      </form>
    );
  }

  render() {
    if (this.props.voucherId !== null) {
      return this.renderRemove();
    }

    return this.renderAdd();
  }
}

CouponForm.propTypes = {
  loading: PropTypes.bool,
  code: PropTypes.string,
  voucherId: PropTypes.number,
  error: PropTypes.string,
  addCoupon: PropTypes.func.isRequired,
  removeCoupon: PropTypes.func.isRequired,
  updateCouponDraft: PropTypes.func.isRequired,
  intl: intlShape.isRequired,
  benefit: PropTypes.shape({
    type: PropTypes.oneOf([PERCENTAGE_BENEFIT, ABSOLUTE_BENEFIT]).isRequired,
    value: PropTypes.number.isRequired,
  }),
};

CouponForm.defaultProps = {
  loading: false,
  code: '',
  voucherId: null,
  error: null,
  benefit: null,
};

const mapStateToProps = state => state.payment.coupon;

export default connect(
  mapStateToProps,
  {
    updateCouponDraft,
    addCoupon,
    removeCoupon,
  },
)(injectIntl(CouponForm));
