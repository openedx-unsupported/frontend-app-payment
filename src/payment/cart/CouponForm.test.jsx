import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import defaultsDeep from 'lodash.defaultsdeep';
import configureMockStore from 'redux-mock-store';
import { createSerializer } from 'enzyme-to-json';
import { IntlProvider } from '@edx/frontend-platform/i18n';
import { sendTrackEvent } from '@edx/frontend-platform/analytics';

import CouponForm from './CouponForm';
import { addCoupon, removeCoupon } from '../data/actions';

// run enzyme JSON serializer using options compatible with prior snapshots
expect.addSnapshotSerializer(createSerializer({ mode: 'deep', noKey: true }));

jest.mock('@edx/frontend-platform/analytics', () => ({
  sendTrackEvent: jest.fn(),
}));

const mockStore = configureMockStore();

const baseState = {
  payment: {
    basket: {
      coupons: [],
      isBasketProcessing: false,
    },
  },
};

const baseCoupon = {
  id: 1234,
  code: 'COUPONCODE',
  benefitValue: '10',
  benefitType: 'Absolute',
};

const renderCouponFormWithState = store => (
  <IntlProvider locale="en">
    <Provider store={store}>
      <CouponForm />
    </Provider>
  </IntlProvider>
);

describe('CouponForm', () => {
  it('should render an absolute value coupon', () => {
    const component = renderCouponFormWithState(mockStore(defaultsDeep({
      payment: {
        basket: {
          coupons: [{ ...baseCoupon, benefitType: 'Absolute' }],
        },
      },
    }, baseState)));
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render a percentage value coupon', () => {
    const component = renderCouponFormWithState(mockStore(defaultsDeep({
      payment: {
        basket: {
          coupons: [{ ...baseCoupon, benefitType: 'Percentage' }],
        },
      },
    }, baseState)));
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render a coupon even if no known benefitType is supplied', () => {
    const component = renderCouponFormWithState(mockStore(defaultsDeep({
      payment: {
        basket: {
          coupons: [{ ...baseCoupon, benefitType: 'Barter' }],
        },
      },
    }, baseState)));
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render a form when there is no coupon', () => {
    const component = renderCouponFormWithState(mockStore(baseState));
    const tree = mount(component);
    expect(tree).toMatchSnapshot();
  });

  describe('Add Coupon', () => {
    const store = mockStore(baseState);
    const component = renderCouponFormWithState(store);
    const wrapper = mount(component);

    it('sends a track event on clicking the add submit coupon button', () => {
      wrapper.find('button[type="submit"]').hostNodes().simulate('click');

      expect(sendTrackEvent).toHaveBeenCalledWith(
        'edx.bi.ecommerce.basket.voucher_applied',
        { type: 'click', category: 'voucher-application' },
      );
    });

    it('fires an add coupon action on submit', () => {
      wrapper.find('form').hostNodes().simulate('submit', {
        target: { elements: { couponField: { value: 'testvalue' } } },
      });

      expect(store.getActions()).toEqual([
        addCoupon({ code: 'testvalue' }),
      ]);
    });
  });

  describe('Remove Coupon', () => {
    const store = mockStore(defaultsDeep({
      payment: {
        basket: {
          coupons: [baseCoupon],
        },
      },
    }, baseState));
    const component = renderCouponFormWithState(store);
    const wrapper = mount(component);

    it('fires an remove coupon action on submit', () => {
      wrapper.find('form').hostNodes().simulate('submit');

      expect(store.getActions()).toEqual([
        removeCoupon({ id: 1234 }),
      ]);
    });
  });
});
