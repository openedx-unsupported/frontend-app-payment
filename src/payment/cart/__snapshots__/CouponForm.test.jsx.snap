// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CouponForm should render a coupon even if no known benefitType is supplied 1`] = `
<form
  className="d-flex align-items-center mb-3"
  onSubmit={[Function]}
>
  <span
    className="text-muted"
  >
    Coupon COUPONCODE applied
  </span>
  <button
    className="display-inline p-0 pl-3 border-0 btn btn-link"
    disabled={false}
    type="submit"
  >
    Remove
  </button>
</form>
`;

exports[`CouponForm should render a form when there is no coupon 1`] = `
<form
  className="summary-row d-flex align-items-end"
  onSubmit={[Function]}
>
  <div
    className="pgn__form-group mb-0 mr-2"
    invalid={false}
  >
    Array [
      <label
        className="h6 d-block"
        htmlFor="couponField"
      >
        Add coupon code (optional)
      </label>,
      <div
        className="pgn__form-control-decorator-group"
      >
        <input
          className="form-control"
          defaultValue=""
          id="couponField"
          name="couponField"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
        />
      </div>,
    ]
  </div>
  <button
    className="btn btn-outline-primary"
    disabled={false}
    onClick={[Function]}
    type="submit"
  >
    Apply
  </button>
</form>
`;

exports[`CouponForm should render a percentage value coupon 1`] = `
<form
  className="d-flex align-items-center mb-3"
  onSubmit={[Function]}
>
  <span
    className="text-muted"
  >
    Coupon COUPONCODE applied for 10% off
  </span>
  <button
    className="display-inline p-0 pl-3 border-0 btn btn-link"
    disabled={false}
    type="submit"
  >
    Remove
  </button>
</form>
`;

exports[`CouponForm should render an absolute value coupon 1`] = `
<form
  className="d-flex align-items-center mb-3"
  onSubmit={[Function]}
>
  <span
    className="text-muted"
  >
    Coupon COUPONCODE applied for 
    $10.00
     off
  </span>
  <button
    className="display-inline p-0 pl-3 border-0 btn btn-link"
    disabled={false}
    type="submit"
  >
    Remove
  </button>
</form>
`;
